File: src/oop/Object/AutoDestructable.c
#include "AutoDestructable.r"

void *deleteAllocationAddressNodeFromAllocationTable(
        Legacy_AllocationTable *OBJECT_ALLOCATION_TABLE,
        void *                  allocatedAddress) {
    return OBJECT_ALLOCATION_TABLE->allocationAddressList
            ->deleteNodeThatHasTheGivenData(
                    OBJECT_ALLOCATION_TABLE->allocationAddressList,
                    allocatedAddress);
}

Legacy_ObjectContainer *
deleteAllocationAddressIfNeeded(AutoDestructable *autoDestructable) {
    if (autoDestructable == NULL) { return NULL; }

    Legacy_ObjectContainer *objectContainer = autoDestructable->allocatedAddress;

    if (objectContainer->object->deleteFromAllocationTableInvocationStatus ==
        WAS_NOT_INVOKED) {
        objectContainer->object->deleteFromAllocationTableInvocationStatus =
                WAS_INVOKED_ONCE;

        Legacy_ObjectContainer *allocatedAddressReturnValue =
                deleteAllocationAddressNodeFromAllocationTable(
                        autoDestructable->OBJECT_ALLOCATION_TABLE,
                        autoDestructable->allocatedAddress);

        if (allocatedAddressReturnValue == NULL) {

            /*
             * The address was already deleted from
             * `autoDestructable->OBJECT_ALLOCATION_TABLE`, and thus had probably already
             * been freed too.
             */
            return NULL;
        }
    }

    return objectContainer;
}

void destructAllocatedAddressUnsafe(AutoDestructable *autoDestructable) {
    if (autoDestructable == NULL) { return; }

    free(autoDestructable->object);
    free(autoDestructable);
}

Legacy_ObjectContainer *
AutoDestructableDestructor(AutoDestructable *autoDestructable) {
    if (autoDestructable == NULL) { return NULL; }

    if (autoDestructable->allocatedAddress->object
                ->destructorInvocationStatus == WAS_NOT_INVOKED) {
        autoDestructable->object->destructorInvocationStatus ==
                WAS_INVOKED_ONCE;

        deleteAllocationAddressIfNeeded(autoDestructable);
    }

    if (autoDestructable->object->destructorInvocationStatus ==
        WAS_NOT_INVOKED) {
        autoDestructable->object->destructorInvocationStatus ==
                WAS_INVOKED_ONCE;

        destructAllocatedAddressUnsafe(autoDestructable);
    }
    return NULL;
}

void constructor_AutoDestructable_fields(AutoDestructable *autoDestructable) {
    autoDestructable->object =
            Legacy_ObjectConstructorClassName("AutoDestructable");

    static Constructable const constructable = {
            .constructor =
                    (void *(*const)(void) )(&AutoDestructableConstructor)};
    autoDestructable->object->constructable = &constructable;

    static Destructable const destructable = {
            .destructor =
                    (void *(*const)(void *) )(&AutoDestructableDestructor)};
    autoDestructable->object->destructable = &destructable;

    autoDestructable->object->destructorInvocationStatus = WAS_NOT_INVOKED;
    autoDestructable->object->deleteFromAllocationTableInvocationStatus =
            WAS_NOT_INVOKED;
}

void saveObjectContainerToAllocationTable(AutoDestructable *autoDestructable) {
    autoDestructable->OBJECT_ALLOCATION_TABLE =
            getLegacy_AllocationTableList()
                    ->findLegacy_AllocationTableByClassName(
                            autoDestructable->object->CLASS_NAME);
    if (autoDestructable->OBJECT_ALLOCATION_TABLE == NULL) {
        autoDestructable->OBJECT_ALLOCATION_TABLE =
                Legacy_AllocationTableConstructorWithClassName(
                        (char *) autoDestructable->object->CLASS_NAME);

        // Create a legacy_node that its data points to `autoDestructable->OBJECT_ALLOCATION_TABLE`.
        Legacy_Node *nodeThatItsDataPointsClassAllocationTable =
                Legacy_NodeConstructorWithData(
                        autoDestructable->OBJECT_ALLOCATION_TABLE);

        // Add this legacy_node to `GLOBAL_ALLOCATION_TABLE_LIST->legacy_allocationTableList`.
        getLegacy_AllocationTableList()->allocationTableList->addAsUnique(
                getLegacy_AllocationTableList()->allocationTableList,
                nodeThatItsDataPointsClassAllocationTable,
                getLegacy_AllocationTableList()
                        ->predicateFindLegacy_AllocationTableByClassName,
                autoDestructable->object->CLASS_NAME);
    }

    // Create a legacy_node that its data points to the "pointer of `autoDestructable->allocatedAddress`".
    Legacy_Node *nodeThatItsDataPointsToThePointerOfObj =
            Legacy_NodeConstructorWithData(autoDestructable->allocatedAddress);

    // Add this legacy_node to `autoDestructable->OBJECT_ALLOCATION_TABLE->allocationAddressList`.
    autoDestructable->OBJECT_ALLOCATION_TABLE->allocationAddressList->add(
            autoDestructable->OBJECT_ALLOCATION_TABLE->allocationAddressList,
            nodeThatItsDataPointsToThePointerOfObj);
}

AutoDestructable *AutoDestructableConstructorWithClassName(
        Legacy_ObjectContainer
                *objectContainerToSaveItsAddressToAllocationTable,
        const char *     className) {
    AutoDestructable *instance = calloc(1, sizeof *instance);
    if (instance == NULL) { /* error handling here */
    }

    constructor_AutoDestructable_fields(instance);

    instance->object->CLASS_NAME = className;

    // If `objectContainerToSaveItsAddressToAllocationTable` is `NULL` use `instance`.
    instance->allocatedAddress =
            objectContainerToSaveItsAddressToAllocationTable == NULL
                    ? (Legacy_ObjectContainer *) instance
                    : objectContainerToSaveItsAddressToAllocationTable;
    saveObjectContainerToAllocationTable(instance);

    return instance;
}

AutoDestructable *AutoDestructableConstructor() {
    return AutoDestructableConstructorWithClassName(NULL, "AutoDestructable");
}

File: src/oop/Object/AutoDestructable.r
#ifndef AUTODESTRUCTABLE_H
#define AUTODESTRUCTABLE_H

#include "Constructable.r"
#include "Destructable.r"
#include "InvocationStatus.r"
#include "Legacy_AllocationTable.r"
#include "Legacy_AllocationTableList.r"
#include "Legacy_ObjectComponent.r"
#include "Legacy_ObjectContainer.r"
#include <stdio.h>
#include <stdlib.h>

// Forward declaration of incomplete type
typedef struct autoDestructable AutoDestructable;

struct autoDestructable {

    /// `AutoDestructable` implements `Legacy_ObjectComponent`.
    Legacy_ObjectComponent *legacyObjectComponent;

    /// Singleton for the whole autoDestructable. Sensitive data. DO NOT TOUCH!
    Legacy_AllocationTable *OBJECT_ALLOCATION_TABLE;

    /// Sensitive data. DO NOT TOUCH!
    Legacy_Object *allocatedAddress;
};

AutoDestructable *AutoDestructableConstructorWithClassName(
        Legacy_Object *legacyObjectToSaveItsAddressToAllocationTable,
        const char *className);

/**
 * @brief Creates a `new` "heap-allocated" instance of `AutoDestructable` and initializes
 *        it.
 *
 * @attention Remember to `free` the returned value of this function.
 *
 * For example:
 * @code
 *
 * // Construct.
 * AutoDestructable *autoDestructable = AutoDestructableConstructor();
 *
 * // Do something...
 * (*autoDestructable->print)(autoDestructable);
 * (*autoDestructable->addOneToX)(autoDestructable);
 * (*autoDestructable->print)(autoDestructable);
 *
 * // Remember to `free`.
 * free(autoDestructable);
 * @endcode
 *
 * @return A pointer to a `new` "heap-allocated" instance of `AutoDestructable`.
 */
AutoDestructable *AutoDestructableConstructor();

Legacy_Object *
AutoDestructableDestructor(AutoDestructable *autoDestructable);

#endif //AUTODESTRUCTABLE_H
